queue.go

```diff
@@ -0,0 +1,423 @@
+			if hostname == "" {
+				hostname = ctx.OurHostname
+			}
+
+			if parts[1] == hostname {
+				log.Debugf("remote_only, skipping %s", rcpt)
+				continue
+			}
+		}
+
+		newRcpts = append(newRcpts, rcpt)
+	}
+
+	ctx.To = newRcpts
+	return nil
+}
+
+func (q *Queue) Deliver(ctx module.DeliveryContext, msg io.Reader) error {
+	if err := filterRcpts(&ctx); err != nil {
+		return err
+	}
+
+	if len(ctx.To) == 0 {
+		return nil
+	}
+
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return fmt.Errorf("message UUID generation failed: %v", err)
+	}
+
+	var body io.ReadSeeker
+	if seekable, ok := msg.(io.ReadSeeker); ok {
+		body = seekable
+	} else {
+		bodySlice, err := ioutil.ReadAll(msg)
+		if err != nil {
+			return errors.New("failed to buffer message")
+		}
+		body = bytes.NewReader(bodySlice)
+	}
+
+	go func() {
+		q.Log.Printf("accepted message %s from %s (%s)", id, ctx.SrcAddr, ctx.SrcHostname)
+		meta := &QueueMetadata{
+			UUID:        id,
+			TriesCount:  1,
+			Ctx:         &ctx,
+			LastAttempt: time.Now(),
+		}
+		shouldRetry, newRcpts := q.attemptDelivery(meta, body)
+		if !shouldRetry {
+			return
+		}
+
+		meta.Ctx.To = newRcpts
+		if _, err := body.Seek(0, io.SeekStart); err != nil {
+			q.Log.Printf("failed to rewind message buffer: %v", err)
+		}
+		if err := q.storeNewMessage(meta, body); err != nil {
+			q.Log.Printf("failed to save message %v to disk: %v", id, err)
+			return
+		}
+		q.Log.Printf("will retry in %v (at %v)", q.initialRetryTime, time.Now().Add(q.initialRetryTime))
+		q.wheel.Add(time.Now().Add(q.initialRetryTime), meta.UUID)
+	}()
+
+	return nil
+}
+
+func (q *Queue) removeFromDisk(id uuid.UUID) {
+	bodyPath := filepath.Join(q.location, id.String())
+	if err := os.Remove(bodyPath); err != nil {
+		q.Log.Printf("failed to remove %s body from disk", id)
+	}
+	metaPath := bodyPath + ".meta"
+	if err := os.Remove(metaPath); err != nil {
+		q.Log.Printf("failed to remove %s meta-data from disk", id)
+	}
+}
+
+func (q *Queue) readDiskQueue() error {
+	dirInfo, err := ioutil.ReadDir(q.location)
+	if err != nil {
+		return err
+	}
+
+	loadedCount := 0
+	for _, entry := range dirInfo {
+		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".meta") {
+			continue
+		}
+		bodyName := entry.Name()[:len(entry.Name())-5]
+		if _, err := os.Stat(filepath.Join(q.location, bodyName)); err != nil {
+			q.Log.Printf("failed to stat body file %s, removing meta-data file", bodyName)
+			os.Remove(entry.Name())
+			continue
+		}
+
+		id, err := uuid.Parse(bodyName)
+		if err != nil {
+			q.Log.Printf("%s is not a valid UUID, skipping", bodyName)
+			continue
+		}
+
+		meta, err := q.readMessageMeta(id)
+		if err != nil {
+			q.Log.Printf("failed to read meta-data for %s, skipping: %v", bodyName, err)
+			continue
+		}
+
+		nextTryTime := meta.LastAttempt
+		nextTryTime = nextTryTime.Add(q.initialRetryTime + q.initialRetryTime*time.Duration(math.Pow(q.retryTimeScale, float64(meta.TriesCount-1))))
```

Comment:
Hmm, the first time this will be `q.initialRetryTime + q.initialRetryTime` instead of `q.initialRetryTime` right?

Reply:
Woops.

Reply:
Yeah, it should be just q.initialRetryTime. ce1b726